packages/go_router/lib/src/shell_state.dart

```diff
@@ -0,0 +1,192 @@
+import 'package:flutter/widgets.dart';
+
+import '../go_router.dart';
+import 'misc/stateful_navigation_shell.dart';
+
+/// The snapshot of the current state of a [StatefulShellRoute].
+///
+/// Note that this an immutable class, that represents the snapshot of the state
+/// of a StatefulShellRoute at a given point in time. Therefore, instances of
+/// this object should not be stored, but instead fetched fresh when needed,
+/// using the method [StatefulShellRouteState.of].
+@immutable
+class StatefulShellRouteState {
+  /// Constructs a [StatefulShellRouteState].
+  const StatefulShellRouteState({
+    required this.route,
+    required this.branchStates,
+    required this.currentIndex,
+    required void Function(StatefulShellBranchState, bool) switchActiveBranch,
+    required void Function(StatefulShellBranchState? branchState,
+            bool navigateToDefaultLocation)
+        resetState,
+  })  : _switchActiveBranch = switchActiveBranch,
+        _resetState = resetState;
+
+  /// Constructs a copy of this [StatefulShellRouteState], with updated values
+  /// for some of the fields.
+  StatefulShellRouteState copy(
+      {List<StatefulShellBranchState>? branchStates, int? currentIndex}) {
+    return StatefulShellRouteState(
+      route: route,
+      branchStates: branchStates ?? this.branchStates,
+      currentIndex: currentIndex ?? this.currentIndex,
+      switchActiveBranch: _switchActiveBranch,
+      resetState: _resetState,
+    );
+  }
+
+  /// The associated [StatefulShellRoute]
+  final StatefulShellRoute route;
+
+  /// The state for all separate route branches associated with a
+  /// [StatefulShellRoute].
+  final List<StatefulShellBranchState> branchStates;
+
+  /// The state associated with the current [StatefulShellBranch].
+  StatefulShellBranchState get currentBranchState => branchStates[currentIndex];
+
+  /// The index of the currently active [StatefulShellBranch].
+  ///
+  /// Corresponds to the index of the branch in the List returned from
+  /// branchBuilder of [StatefulShellRoute].
+  final int currentIndex;
+
+  /// The Navigator key of the current navigator.
+  GlobalKey<NavigatorState> get currentNavigatorKey =>
+      currentBranchState.branch.navigatorKey;
+
+  final void Function(StatefulShellBranchState, bool) _switchActiveBranch;
+
+  final void Function(
+          StatefulShellBranchState? branchState, bool navigateToDefaultLocation)
+      _resetState;
+
+  /// Navigate to the current location of the shell navigator with the provided
+  /// index.
+  ///
+  /// This method will switch the currently active [Navigator] for the
+  /// [StatefulShellRoute] by replacing the current navigation stack with the
+  /// one of the route branch identified by the provided index. If resetLocation
+  /// is true, the branch will be reset to its initial location
+  /// (see [StatefulShellBranch.initialLocation]).
+  void goBranch({
+    required int index,
+    bool resetLocation = false,
```

Comment:
Is there a use case for this? I assume people can use go() with default location directly. Can we remove this boolean?

Reply:
same for all the other `refresh`, `reset` api. I think let's focus on the essential API first.

Reply:
I agree - we'll remove these for now, and perhaps revisit them as refinements in future PRs.